pkg/parse/run.go

```diff
@@ -195,6 +161,51 @@ func DefaultRunFunc(ctx context.Context, r Reconciler, trigger string) RunResult
 	return result
 }
 
+// Fetch waits for the *-sync sidecars to fetch the source manifests to the
+// shared source volume.
+// Updates the RSync status (source status and syncing condition).
+func (r *reconciler) Fetch(ctx context.Context) (*SourceStatus, cmpath.Absolute, status.MultiError) {
+	opts := r.Options()
+	state := r.ReconcilerState()
+	var syncPath cmpath.Absolute
+	newSourceStatus := &SourceStatus{}
+
+	// pull the source commit and directory with retries within 5 minutes.
+	newSourceStatus.Commit, syncPath, newSourceStatus.Errs = hydrate.SourceCommitAndSyncPathWithRetry(
+		util.SourceRetryBackoff, opts.SourceType, opts.SourceDir, opts.SyncDir, opts.ReconcilerName)
+
+	// Add pre-sync annotations to the object.
+	// If updating the object fails, it's likely due to a signature verification error
+	// from the webhook. In this case, add the error as a source error.
+	if newSourceStatus.Errs == nil {
+		if err := r.SyncStatusClient().SetImageToSyncAnnotation(ctx, newSourceStatus.Commit); err != nil {
+			newSourceStatus.Errs = status.Append(newSourceStatus.Errs, err)
+		}
+	}
+
+	// Generate source spec from Reconciler config
+	newSourceStatus.Spec = SourceSpecFromFileSource(opts.FileSource, opts.SourceType, newSourceStatus.Commit)
+
+	// Only update the source status if there are errors or the commit changed.
+	// Otherwise, parsing errors may be overwritten.
+	// TODO: Decouple fetch & parse stages to use different status fields
+	if newSourceStatus.Errs != nil || state.status.SourceStatus == nil || newSourceStatus.Commit != state.status.SourceStatus.Commit {
+		newSourceStatus.LastUpdate = nowMeta(opts.Clock)
+		if state.status.needToSetSourceStatus(newSourceStatus) {
+			klog.V(3).Info("Updating source status (after fetch)")
+			if statusErr := r.SyncStatusClient().SetSourceStatus(ctx, newSourceStatus); statusErr != nil {
+				return newSourceStatus, syncPath, status.Append(newSourceStatus.Errs, statusErr)
+			}
+			state.status.SourceStatus = newSourceStatus
+		}
+		// If there were fetch errors, stop, log them, and retry later
+		if newSourceStatus.Errs != nil {
+			return newSourceStatus, syncPath, newSourceStatus.Errs
+		}
+	}
+	return newSourceStatus, syncPath, nil
```

Comment:
nit:
```golang
return newSourceStatus, syncPath, newSourceStatus.Errs
```
Then the condition on line 202 can be removed

Reply:
yeah, but I was deliberately trying to exit early on errors. It makes it easier to read and easier to extend later without needing to undo the optimization. Like, if I wanted to log when successful, or wrap the error with more context, for example.

Reply:
Added "X Successful" comments to illustrate.